refactor: fix IP parse

x1ah opened the pull request:
- Upgrade Lura to v2
- Fix gin ip parse, related issue: https://github.com/gin-gonic/gin/issues/2697 , close #3 , #4 and #5 
- bug fix

x1ah committed:
refactor: fix gin ip parse

x1ah committed:
feat: add mode args, close #4

x1ah committed:
fix: fix readme

x1ah committed:
chore: fix ci

gfreezy commented:
`allow` should be prohibited in `allow_mode`, and `deny` show be prohibited in `deny mode`. We should ensure this while parsing config.

cadl commented:
Thank you @x1ah, this is a very valuable bugfix.

Can we remove the 'mode' parameter to avoid confusion, such as the case @gfreezy  mentioned with 'allow' and 'mode=allow_all'? This parameter can be easily misunderstood.

After removing the 'mode' parameter, we can continue to use the current black-and-white list mode and restrict the use of 'allow' and 'deny' parameters so that only one can be used, similar to the Apisix ip-restriction plugin (https://apisix.apache.org/docs/apisix/plugins/ip-restriction/).

Alternatively, we can also refer to the Kong ip-restriction plugin, where both 'allow' and 'deny' configurations can be used simultaneously. 
> An interesting use case of this flexibility is to allow a CIDR range but deny an IP address on that CIDR range (https://docs.konghq.com/hub/kong-inc/ip-restriction/#usage).

If you are short on time, I am happy to continue building on what you have provided.

x1ah committed:
- fix raw ip parse error
- remove mode parameter
- consistent with the Kong Gateway IP restriction plugin

x1ah committed:
chore: update readme

x1ah committed:
fix: fix ci

x1ah commented:
@cadl I just made some changes and now it's consistent with Kong's strategy

- Kong IP Restriction docs: https://docs.konghq.com/hub/kong-inc/ip-restriction/
- Kong implements: https://github.com/Kong/kong/blob/master/kong/plugins/ip-restriction/handler.lua#L67
